File: src/benchmark/trading-market.mjs
// One explicit draft broker fixture, independent of hidden strategy rules.
// Existing tickets fill before the completed-bar callback. A new submission
// processes prior cancellation requests; otherwise they settle after the bar.
// A new order cannot fill on its submission bar.
import { canonical, invariant, object } from './prompts.mjs'
import { createTradingRuntime } from './trading-runtime.mjs'
import { validateTradingIR } from './trading-ir.mjs'
import { observeTradingArtifacts, tradingObservationContract, TRADING_TAG_VERSION } from './trading-observations.mjs'

export function validateTradingMarket(ir, input) {
  validateTradingIR(ir)
  invariant(object(input) && Object.keys(input).every(key => ['execution', 'market', 'bars'].includes(key)), 'Declare operational execution, market and completed bars.')
  const market = input.market
  invariant(object(market) && Object.keys(market).every(key => ['version', 'kind', 'delayBars', 'maxFillQuantity'].includes(key))
    && market.version === 1 && market.kind === 'delayed-capped-equity'
    && Number.isSafeInteger(market.delayBars) && market.delayBars >= 1 && market.delayBars <= 10000
    && Number.isSafeInteger(market.maxFillQuantity) && market.maxFillQuantity >= 1 && market.maxFillQuantity <= 1000000, 'Declare the bounded delayed/capped equity fill fixture.')
  invariant(ir.execution.currency === 'USD' && ir.execution.cashCents > 0 && ir.execution.assets.every(asset => /^[A-Z][A-Z0-9.]{0,15}$/.test(asset.id) && asset.multiplier === 1 && asset.quantityStep === 1), 'This profile requires positive USD cash and unit whole-share equity assets.')
  invariant(Array.isArray(input.bars) && input.bars.length > 0 && input.bars.length <= Math.min(100000, ir.contract.barBudget), 'Declare the complete bounded market bar input.')
  const assets = new Set(ir.execution.assets.map(asset => asset.id))
  let previous = -1
  for (const bar of input.bars) {
    invariant(object(bar) && Object.keys(bar).every(key => ['time', 'prices'].includes(key)) && Number.isSafeInteger(bar.time) && bar.time >= 0 && bar.time <= 4102444800 && bar.time % 60 === 0 && bar.time > previous && object(bar.prices), 'Operational bars require ordered minute-aligned UTC seconds.')
    invariant(Object.entries(bar.prices).every(([asset, price]) => assets.has(asset) && (price === null || Number.isSafeInteger(price) && price > 0)) && Object.values(bar.prices).some(price => price !== null), 'An operational bar needs observed declared equity prices.')
    previous = bar.time
  }
  return input
}

export function simulateTradingMarket(ir, input) {
  validateTradingMarket(ir, input)
  const orders = {}, events = [], pending = new Map(), cancellations = [], paths = new Map(ir.nodes.map(node => [node.namespace, node.path]))
  let index = -1, serial = 0, runtime
  const emit = (ticket, status, quantity = 0, price = 0) => {
    const row = { orderId: ticket.native, orderEventId: ++ticket.serial, time: input.bars[index].time, status, symbolValue: ticket.intent.asset,
      quantity: ticket.order.quantity, fillQuantity: quantity, fillPrice: price / 100, fillPriceCurrency: 'USD', orderFeeAmount: 0, orderFeeCurrency: 'USD' }
    events.push(row); ticket.order.status = status
    if (status === 'cancelPending') return
    runtime.reconcile({ id: ticket.native + ':' + ticket.serial, intentId: ticket.intent.id, brokerOrderId: String(ticket.native), time: row.time,
      kind: ({ submitted: 'accepted', filled: 'fill', partiallyFilled: 'fill', canceled: 'cancelled' })[status],
      ...(quantity ? { quantity: Math.abs(quantity), priceCents: price, feeCents: 0 } : {}) })
  }
  const settleCancellations = () => {
    for (const id of cancellations.splice(0)) {
      const ticket = pending.get(id)
      invariant(ticket, 'A cancellation lost its native ticket before acknowledgment.')
      emit(ticket, 'canceled'); pending.delete(id)
    }
  }
  runtime = createTradingRuntime(ir, {
    dispatch(intent) {
      settleCancellations()
      const native = ++serial, order = { id: native, type: 0, securityType: 1, symbol: { value: intent.asset }, priceCurrency: 'USD',
        time: intent.time, quantity: intent.quantity * (intent.side === 'buy' ? 1 : -1), status: 'new',
        tag: canonical([TRADING_TAG_VERSION, paths.get(intent.owner), intent.lotId, intent.reason, intent.id]) }
      const ticket = { native, order, intent, created: index, serial: 0, filled: 0 }
      orders[native] = order; pending.set(intent.id, ticket); emit(ticket, 'submitted')
    },
    cancel(id) {
      invariant(pending.has(id), 'Simulation cancellation has no pending native order.')
      emit(pending.get(id), 'cancelPending'); cancellations.push(id)
    },
  })
  for (const bar of input.bars) {
    index++
    for (const [id, ticket] of pending) {
      const price = bar.prices[ticket.intent.asset]
      if (index - ticket.created < input.market.delayBars || price === null || price === undefined) continue
      const quantity = Math.min(input.market.maxFillQuantity, ticket.intent.quantity - ticket.filled)
      ticket.filled += quantity
      emit(ticket, ticket.filled === ticket.intent.quantity ? 'filled' : 'partiallyFilled', quantity * (ticket.intent.side === 'buy' ? 1 : -1), price)
      if (ticket.filled === ticket.intent.quantity) pending.delete(id)
    }
    runtime.step(bar)
    settleCancellations()
  }
  return { observation: observeTradingArtifacts(tradingObservationContract(ir, input), orders, events), snapshot: runtime.snapshot() }
}

File: src/benchmark/trading_market.py
"""Independent Python broker simulation for the declared draft study fixture.

This interpreter neither imports JavaScript nor executes candidate code. It
produces its own normalized orders, receipt sequence and private fill book.
Native qualification must additionally compare the actual LEAN artifacts.
"""
import json
import sys
from trading_reference import TradingRuntime


def simulate(ir, data):
    orders, events, lots, pending, cancellations = [], [], [], {}, []
    owners = {node["namespace"]: node["path"] for node in ir["nodes"]}
    local_lots, last_prices = {}, {}
    cash = ir["execution"]["cashCents"]
    positions = {asset["id"]: 0 for asset in ir["execution"]["assets"]}
    index = -1

    def receipt(ticket, status, quantity=0, price=0):
        nonlocal cash
        order = ticket["order"]
        ticket["serial"] += 1
        now = data["bars"][index]["time"]
        order["status"] = status
        order["pending"] = status not in ("filled", "cancelled", "rejected")
        events.append(dict(order=order["order"], time=now, status=status,
                           quantity=quantity, priceCents=price, feeCents=0))
        if quantity:
            order["filledQuantity"] += abs(quantity)
            order["unfilledQuantity"] -= abs(quantity)
            lot = ticket["lot"]
            if quantity > 0:
                lot["boughtQuantity"] += quantity
                if lot["firstFillTime"] is None:
                    lot["firstFillTime"] = now
            else:
                lot["soldQuantity"] -= quantity
            lot["quantity"] += quantity
            lot["lastFillTime"] = now
            positions[order["asset"]] += quantity
            cash -= quantity * price
        if status == "cancel-pending":
            return
        event = dict(id=str(order["order"]) + ":" + str(ticket["serial"]),
                     intentId=ticket["intent"]["id"], brokerOrderId=str(order["order"]), time=now,
                     kind={"accepted": "accepted", "filled": "fill", "partial": "fill", "cancelled": "cancelled"}[status])
        if quantity:
            event.update(quantity=abs(quantity), priceCents=price, feeCents=0)
        runtime.reconcile(event)

    def submit(intent):
        settle_cancellations()
        owner = owners[intent["owner"]]
        local = local_lots.setdefault(owner, {})
        if intent["side"] == "buy":
            lot = dict(owner=owner, lot=len(local) + 1, asset=intent["asset"], boughtQuantity=0,
                       soldQuantity=0, quantity=0, firstFillTime=None, lastFillTime=None)
            local[intent["lotId"]] = lot
            lots.append(lot)
        lot = local[intent["lotId"]]
        order = dict(order=len(orders) + 1, owner=owner, lot=lot["lot"], asset=intent["asset"], time=intent["time"],
                     quantity=intent["quantity"] * (1 if intent["side"] == "buy" else -1), reason=intent["reason"],
                     status="new", filledQuantity=0, feesCents=0, unfilledQuantity=intent["quantity"], pending=True)
        orders.append(order)
        ticket = dict(intent=intent, order=order, lot=lot, created=index, serial=0)
        pending[intent["id"]] = ticket
        receipt(ticket, "accepted")

    def cancel(identifier):
        receipt(pending[identifier], "cancel-pending")
        cancellations.append(identifier)

    def settle_cancellations():
        for identifier in cancellations:
            receipt(pending.pop(identifier), "cancelled")
        cancellations.clear()

    runtime = TradingRuntime(ir, dispatch=submit, cancel=cancel)
    for index, bar in enumerate(data["bars"]):
        last_prices.update({asset: price for asset, price in bar["prices"].items() if price is not None})
        for identifier, ticket in list(pending.items()):
            price = bar["prices"].get(ticket["intent"]["asset"])
            if index - ticket["created"] < data["market"]["delayBars"] or price is None:
                continue
            quantity = min(data["market"]["maxFillQuantity"], ticket["order"]["unfilledQuantity"])
            filled = quantity == ticket["order"]["unfilledQuantity"]
            receipt(ticket, "filled" if filled else "partial", quantity * (1 if ticket["intent"]["side"] == "buy" else -1), price)
            if filled:
                del pending[identifier]
        runtime.step(bar)
        settle_cancellations()
    observation = dict(format="lean-operational-observation", version=1, orders=orders, events=events, lots=lots,
                       cashFromFillsCents=cash, positionsFromFills=[dict(asset=asset, quantity=quantity) for asset, quantity in positions.items()],
                       feesCents=0, equityCents=dict(start=ir["execution"]["cashCents"],
                                                  end=cash + sum(quantity * last_prices.get(asset, 0) for asset, quantity in positions.items())))
    return dict(observation=observation, snapshot=runtime.snapshot())


if __name__ == "__main__":
    task = json.load(sys.stdin)
    print(json.dumps(simulate(task["ir"], task["input"]), separators=(",", ":"), allow_nan=False))
